### ISSUE TITLE
Ideas

### ISSUE THREAD

**Issue by jtwatson** [alanctgardner/gogen-avro#30]
Hello,

I just wanted to bring up a few things that I ran across when using your lib in my project.  I am able to make things work as-is, but thought it might be helpful to share with you.

First, my project is rather large.  I am writing a program to extract a over 1000 sql views out of one product and dump them into another product.  The destination product accepts avro files, so this is the intermediate format that I have chosen.  So thank you for your work!  :)

I wrote a program to generate go structs for each view being extracted.  These structs are then used with the Go SQL package to extract data.  I then generate avro schema files and use gogen-avro to generate go code that can be used to export the data to avro container files.  So, I execute gogen-avro with **1300 avro schema files on the command line** to generate the code.

**Issues:** 
1. The comment block that is inserted into almost every generated source code file contains a full list of all avro schema files.  So, I get 1300 files which each have a list 1300 source files at the top.  This quickly gets out of hand regarding scalability.
2. It would be really nice if there was a way to invoke your code programmatically.  I could include your code in my code generation project, and pass the avro schema definitions to your code via sice of []byte or string.  This way I would not need to first right the schema files to the file system, and then invoke your command line tool.  Wish list item I suppose, but I think it would be easy to rewrite things a bit to provide a programmatic API that can invoke your code in the same manner as the command line tool.

**Comment by alanctgardner** [comment]
Hey @jtwatson, I'm glad to hear this project has been helpful! I never thought about this kind of use case, but I made some adjustments on master that I think will help. You should be able to programmatically generate your files like this:

```golang
import (
    "github.com/alanctgardner/gogen-avro/types"
    "github.com/alanctgardner/gogen-avro/generator"
)

// Make a new namespace to add all the types to
namespace := types.NewNamespace()
// Call this once for each schema you want to add
_, err = namespace.TypeForSchema(<string with the Avro schema>)

// Add all the structs to the package - don't include any headers
pkg := generator.NewPackage("mypackage")
err = namespace.AddToPackage(pkg, "", false)

// Write the package with all the structs to the output directory
err = pkg.WriteFiles("output-directory")
```

**Comment by jtwatson** [comment]
Thank you very much!  This worked awesome!

**Comment by sbramin** [comment]
@jtwatson I have a very similar use case to yours and am currently trying to work out the best approach to handling nulls from SQL sources.  Initially I thought I could just initialize the struct created by gogen-avro and use it for pulling out data from a rows.Scan() and then pushing that same struct into WriteRecord().  Then for fields with nulls I created a separate var of type sql.NullString etc and used if field.Valid { myStruct.Field.String = field.String}.  This appears to work, especially for strings where the default value of "" looks a bit like a null but is really quite different.  And gets even worse for Ints/Floats where I don't want a 0 value but a true null.

It seems if I use a UnionNullString I just always get null even when the value is populated and if I use UnionStringNull I always get the default value of that type (eg string "", or 0 for int).  The README.md says the first value will be the default, but how do you override said default.

Thanks in advance for any help.



:

**Comment by sbramin** [comment]
Figured out why always null issue, needed to set the UnionType,
`                if pbsp.Valid {
                        x.PLAYBACKSPEED.Long = pbsp.Int64
                        x.PLAYBACKSPEED.UnionType = 1
                }
`
Sometimes its necessary to say something out loud to fully understand it.

**Comment by jtwatson** [comment]
@sbramin I generated separate structs for the source and destination, and then generated a copy() function that would move data from the source struct into the gogen-avro generated destination struct.  In this manner I was able to fully handle null types without any issue at all.  I also use this step for data transformation.

Regarding the performance question, I never use linkedin's avro lib, so I can't compare performance, but I know that my slowest point is the source system.  It is quite slow, so I suspect I wouldn't see much difference.
